qualang_tools/virtual_gates/virtual_gate_sequence.py

```diff
@@ -0,0 +1,252 @@
+                voltage_point_name = "unregistered_value"
+                voltage_level = level[i]
+            else:
+                raise RuntimeError(
+                    "Either the voltage_point_name or the duration and desired voltage level must be provided."
+                )
+            # Play a step
+            if ramp_duration is None:
+                self.average_power[i] += self._update_averaged_power(voltage_level, _duration)
+
+                # Dynamic amplitude change...
+                if self.is_QUA(voltage_level) or self.is_QUA(self.current_level[i]):
+                    # if dynamic duration --> play step and wait
+                    if self.is_QUA(_duration):
+                        play("step" * amp((voltage_level - self.current_level[i]) * 4), gate)
+                        wait((_duration - 16) >> 2, gate)
+                    # if constant duration --> new operation and play(*amp(..))
+                    else:
+                        operation = self._add_op_to_config(
+                            gate,
+                            "step",
+                            amplitude=0.25,
+                            length=_duration,
+                        )
+                        play(operation * amp((voltage_level - self.current_level[i]) * 4), gate)
+
+                # Fixed amplitude but dynamic duration --> new operation and play(duration=..)
+                elif isinstance(_duration, (QuaVariable, QuaExpression)):
+                    operation = self._add_op_to_config(
+                        gate,
+                        voltage_point_name,
+                        amplitude=voltage_level - self.current_level[i],
+                        length=16,
+                    )
+                    play(operation, gate, duration=_duration >> 2)
+
+                # Fixed amplitude and duration --> new operation and play()
+                else:
+                    operation = self._add_op_to_config(
+                        gate,
+                        voltage_point_name,
+                        amplitude=voltage_level - self.current_level[i],
+                        length=_duration,
+                    )
+                    play(operation, gate)
+
+            # Play a ramp
+            else:
+                self.average_power[i] += self._update_averaged_power(
+                    voltage_level, _duration, ramp_duration, self.current_level[i]
+                )
+
+                if not self.is_QUA(ramp_duration):
+                    ramp_rate = 1 / ramp_duration
+                    play(ramp((voltage_level - self.current_level[i]) * ramp_rate), gate, duration=ramp_duration >> 2)
+                    wait(_duration >> 2, gate)
+
+            self.current_level[i] = voltage_level
+
+    def add_compensation_pulse(self, duration: int) -> None:
+        """Add a compensation pulse of the specified duration whose amplitude is derived from the previous operations.
```

Comment:
I guess that this would only work when everything is in strict timing?

Reply:
Added a note in README please check